components/script/dom/underlyingsourcecontainer.rs

```diff
@@ -0,0 +1,138 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use std::rc::Rc;
+
+use dom_struct::dom_struct;
+use js::jsval::UndefinedValue;
+use js::rust::wrappers::JS_GetProperty;
+
+use crate::dom::bindings::callback::ExceptionHandling;
+use crate::dom::bindings::codegen::Bindings::UnderlyingSourceBinding::UnderlyingSource as JsUnderlyingSource;
+use crate::dom::bindings::import::module::UnionTypes::ReadableStreamDefaultControllerOrReadableByteStreamController as Controller;
+use crate::dom::bindings::reflector::{reflect_dom_object_with_proto, DomObject, Reflector};
+use crate::dom::bindings::root::DomRoot;
+use crate::dom::globalscope::GlobalScope;
+use crate::dom::promise::Promise;
+use crate::dom::readablestreamdefaultcontroller::ReadableStreamDefaultController;
+use crate::js::conversions::ToJSValConvertible;
+
+/// <https://streams.spec.whatwg.org/#underlying-source-api>
+#[derive(JSTraceable)]
+pub enum UnderlyingSourceType {
+    /// Facilitate partial integration with sources
+    /// that are currently read into memory.
+    Memory(usize),
+    /// A blob as underlying source, with a known total size.
+    Blob(usize),
+    /// A fetch response as underlying source.
+    FetchResponse,
+    /// A JS object as underlying source.
+    Js(JsUnderlyingSource),
+}
+
+impl UnderlyingSourceType {
+    /// Is the source backed by a Rust native source?
+    pub fn is_native(&self) -> bool {
+        match self {
+            UnderlyingSourceType::Memory(_) |
+            UnderlyingSourceType::Blob(_) |
+            UnderlyingSourceType::FetchResponse => true,
+            _ => false,
+        }
+    }
+
+    /// Does the source have all data in memory?
+    pub fn in_memory(&self) -> bool {
+        matches!(self, UnderlyingSourceType::Memory(_))
+    }
+}
+
+#[dom_struct]
+pub struct UnderlyingSourceContainer {
+    reflector_: Reflector,
+    #[ignore_malloc_size_of = "Rc is hard"]
+    underlying_source_type: UnderlyingSourceType,
+}
+
+impl UnderlyingSourceContainer {
+    fn new_inherited(underlying_source_type: UnderlyingSourceType) -> UnderlyingSourceContainer {
+        UnderlyingSourceContainer {
+            reflector_: Reflector::new(),
+            underlying_source_type: underlying_source_type,
+        }
+    }
+
+    #[allow(unsafe_code)]
+    pub fn new(
+        global: &GlobalScope,
+        underlying_source_type: UnderlyingSourceType,
+    ) -> DomRoot<UnderlyingSourceContainer> {
+        // Setting the prototype of the underlying source dict on the
+        // `UnderlyingSourceContainer` for later use in Call_.
+        // TODO: is this a good idea?
```

Comment:
cc @jdm

Reply:
I ended-up removing this entirely for now, it was crashing.

Reply:
I still think we probably should expand the codegen to support underlying source that can call `Call`. I guess this is because it's the first dictionary that has callbacks.

Reply:
One thing that occurs to me—I believe the DomObject bound us only needed to call wrap_call_this_object, which only uses it to call the reflector method so it can get a JS object pointer. If we add a new trait to return the JS object directly instead, we should not need to play around with dictionary prototypes, just generate an impl for the new trait instead.

Reply:
thank you, have included your example.